### PR TITLE
Use local changes when building and running locally

### PR DESCRIPTION
**Pull request by astencel-sumo**
This is a poor man's try at adding the following improvements:
- When building locally, push the built Docker image to local registry and use it when running locally.
- When building locally, use Docker cache.

Suggestions for improvements are welcome.

### CODE REVIEW

File: ci/build.sh [hunk 1]
```diff
@@ -106,7 +106,9 @@ bundle_fluentd_plugins "${VERSION}" || (echo "Failed bundling fluentd plugins" &
 
 echo "Building docker image with $DOCKER_TAG:local in $(pwd)..."
 cd ./deploy/docker || exit 1
-docker build . -f ./Dockerfile -t "$DOCKER_TAG:local" --no-cache
+readonly no_cache=$([ "$DOCKER_USE_CACHE" == "true" ] && echo "" || echo "--no-cache")
+# shellcheck disable=SC2086
```
Comment (sumo-drosiek) [inline ci/build.sh:110]: what was that?

Comment (sumo-drosiek) [reply ci/build.sh:110]: i think it should be `if [[ ...`

File: vagrant/scripts/generate-local-values-file.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+cat > /sumologic/vagrant/values.local.yaml <<END
+fluentd:
+  metadata:
+    pluginLogLevel: "info"
```
Comment (sumo-drosiek) [inline vagrant/scripts/generate-local-values-file.sh:6]: I doubt we should add this to the auto generated file

File: vagrant/scripts/generate-local-values-file.sh [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+cat > /sumologic/vagrant/values.local.yaml <<END
```
Comment (sumo-drosiek) [inline vagrant/scripts/generate-local-values-file.sh:3]: We shouldn't override existing file

File: vagrant/Makefile [hunk 4]
```diff
@@ -44,10 +57,11 @@ helm-dependencies:
 	helm dependency update "${helm_path}"
 
 helm-upgrade:
+	$(eval local_values_param := $(shell [ -f ${local_values_file} ] && echo "-f ${local_values_file}" || echo ""))
```
Comment (sumo-drosiek) [inline vagrant/Makefile:60]: `eval` and `shell` 🙀

File: vagrant/scripts/generate-local-values-file.sh [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+fluentd:
+  metadata:
+    pluginLogLevel: "info"
+image:
+  repository: localhost:32000/sumologic/kubernetes-fluentd
+  tag: local
+  pullPolicy: Always
+END
```
Comment (sumo-drosiek) [inline vagrant/scripts/generate-local-values-file.sh:11]: Add target for this script to Makefile

**Review by sumo-drosiek — CHANGES_REQUESTED** [review]
See comments

File: vagrant/scripts/generate-local-values-file.sh (hunk 3 above)
Comment (perk-sumo) [reply vagrant/scripts/generate-local-values-file.sh:3]: Hm why not? When you build stuff you override the old build dir.

Comment (sumo-drosiek) [reply vagrant/scripts/generate-local-values-file.sh:3]: I want to keep some configuration for test without need to change it with every build. There would be no difference with using `vagrant/values.yaml` if we keep overriding this file automatically. Anyway this is not happening right now as I see. You have to manually fire the script. Its's not automated anywhere

File: ci/build.sh [hunk 6]
```diff
@@ -106,7 +106,9 @@ bundle_fluentd_plugins "${VERSION}" || (echo "Failed bundling fluentd plugins" &
...
+docker build . -f ./Dockerfile -t "$DOCKER_TAG:local" ${no_cache}
```
Comment (perk-sumo) [inline ci/build.sh:111]: per https://github.com/koalaman/shellcheck/wiki/SC2086#exceptions
```suggestion
docker build . -f ./Dockerfile -t "$DOCKER_TAG:local" ${no_cache:+"no_cache"}
```

File: vagrant/scripts/generate-local-values-file.sh (hunk 2 above)
Comment (perk-sumo) [reply vagrant/scripts/generate-local-values-file.sh:6]: I think it's nice to have developer settings for the development values yaml file.

Comment (sumo-drosiek) [reply vagrant/scripts/generate-local-values-file.sh:6]: I do not agree. every feature development requires another set of the settings and they may conflict with each other

File: vagrant/Makefile (hunk 4 above)
Comment (astencel-sumo) [reply vagrant/Makefile:60]: This seems to be the way to define variables evaluated at rule execution time: https://stackoverflow.com/questions/1909188/define-make-variable-at-rule-execution-time. The Makefile's eval function: https://www.gnu.org/software/make/manual/make.html#Eval-Function

File: ci/build.sh (hunk 1 above)
Comment (astencel-sumo) [reply ci/build.sh]: Done

File: vagrant/scripts/generate-local-values-file.sh (hunk 2 above)
Comment (astencel-sumo) [reply vagrant/scripts/generate-local-values-file.sh]: Removed the `pluginLogLevel` config entry.

File: ci/build.sh (hunk 6 above)
Comment (astencel-sumo) [reply ci/build.sh]: @perk-sumo please see the change

**Review by sumo-drosiek — APPROVED** [review]

**Review by sumo-drosiek — APPROVED** [review]
🐇

**Review by pmalek-sumo — APPROVED** [review]

File: vagrant/Makefile [hunk 7]
```diff
@@ -16,6 +17,21 @@ namespace = sumologic
...
+build-local:
+	DOCKER_USE_CACHE=true \
+	${mkfile_path} \
+		build \
+		tag-and-push-local
+
+tag-and-push-local:
+	$(eval timestamp := $(shell date +%s))
+	docker tag sumologic/kubernetes-fluentd:local localhost:32000/sumologic/kubernetes-fluentd:local-${timestamp}
+	docker push localhost:32000/sumologic/kubernetes-fluentd:local-${timestamp}
```
Comment (perk-sumo) [inline vagrant/Makefile:32]: Can we puch the `local` tag as well?

Comment (pmalek-sumo) [reply vagrant/Makefile:32]: What would be the use case for it?

File: vagrant/Makefile [hunk 8]
```diff
@@ -16,6 +17,21 @@ namespace = sumologic
...
+	sed -i 's/^\([[:space:]]\+\)tag:.*$$/\1tag: local-${timestamp}/' ${local_values_file}
```
Comment (perk-sumo) [inline vagrant/Makefile:33]: Please change to`yq`

**Pull request merged**